postbump-baseline hook sets unpublished tag as PackageValidation baseline

## Summary

`scripts/postbump-baseline.mjs` sets `PackageValidationBaselineVersion` in `Directory.Build.props` to the most recent git tag before the release commit. For v4.1.2, the most recent tag will be `v4.1.1`, which was never published to NuGet.org. When the release workflow restores with `-p:Version=4.1.2`, MSBuild fetches the baseline `QuerySpec.Core.4.1.1` package — which does not exist on NuGet — and the restore step fails with NU1102.

This is the same failure mode that blocked the v3.0.0 release (fixed in PR #116), but in a new form: a valid git tag that was never published.

## Location

- `scripts/postbump-baseline.mjs` — line 21 (`git describe --tags --abbrev=0`)
- `Directory.Build.props` — `PackageValidationBaselineVersion` element

## Evidence

The hook comment on line 8 states: "the previous release on nuget.org — exactly what the next build should validate against." This invariant breaks when a tag exists without a corresponding NuGet publish. Tags `v4.1.0` and `v4.1.1` exist in the repository but neither was published; `v4.0.0` is the last version on NuGet.

Running `git describe --tags --abbrev=0 --match=v*` on the develop branch after `npm run release` bumps to 4.1.2 returns `v4.1.1`. The hook sets `PackageValidationBaselineVersion` to `4.1.1`. CI then fails:

```
error NU1102: Unable to find package QuerySpec.Core with version (= 4.1.1)
  - Found 9 version(s) in nuget.org [ Nearest version: 4.0.0 ]
```

## Proposed remediation

Before running `npm run release` to cut v4.1.2, manually reset `Directory.Build.props` `PackageValidationBaselineVersion` to `4.0.0` (the last published version). The postbump hook's `oldBaseline === prevVersion` idempotency check will see `4.0.0 !== 4.1.1` and overwrite it with `4.1.1` — so the manual reset alone is not sufficient.

The durable fix is to update the hook to verify the tag exists on NuGet before using it as the baseline. A simple `curl -s -o /dev/null -w "%{http_code}" https://api.nuget.org/v3-flatcontainer/queryspec.core/{tag}/queryspec.core.{tag}.nupkg` check (exit 0 on 200, skip to next older tag on 404) makes the hook resilient to unpublished tags.

Alternatively, add a `--baseline-override` argument to the hook so a release manager can explicitly specify the last published version when the standard `git describe` path returns an unpublished tag.

## Risk and verification

Without this fix, the v4.1.2 release will fail at the `Restore` step in CI before any tests or packs run. The risk of the proposed check is a network call from a dev machine at release-prep time; this is acceptable. The acceptance criterion is that `npm run release` on a tree containing only unpublished tags between the current version and the last published version still resolves the correct NuGet baseline.

## Acceptance criteria

- [ ] `scripts/postbump-baseline.mjs` skips git tags that have no corresponding NuGet package, falling back to the nearest older published version
- [ ] Running `npm run release` on a tree with `v4.1.0` and `v4.1.1` tags (both unpublished) produces `PackageValidationBaselineVersion` = `4.0.0`
- [ ] v4.1.2 release CI restore step succeeds

--- a/scripts/postbump-baseline.mjs
+++ b/scripts/postbump-baseline.mjs
@@ -3,33 +3,91 @@
 // PackageValidationBaselineVersion aligned with the previous published release.
 //
 // Lifecycle: standard-version runs this AFTER bumping package.json to the new version
-// but BEFORE creating the release commit and tag. At this point `git describe --tags
-// --abbrev=0 --match=v*` returns the most recent EXISTING tag, which is the previous
-// release on nuget.org — exactly what the next build should validate against.
+// but BEFORE creating the release commit and tag. At this point the most recent v* git
+// tag is the previous release. However that tag may correspond to a release that failed
+// before reaching NuGet (e.g. a CI gate failure). Writing an unpublished version as the
+// baseline causes PackageValidation to fail with NU1102 on the very next release attempt.
+//
+// This script resolves the issue by querying the NuGet v3 flat-container index for each
+// candidate tag (newest first) and stopping at the first one that is actually published.
+// If no candidate is published (brand-new package) it exits 0 without modifying the file.
 //
 // commit-all is enabled in .versionrc.json so the modified Directory.Build.props is
 // included in standard-version's release commit.
 
 import { execSync } from 'node:child_process';
 import { readFileSync, writeFileSync } from 'node:fs';
+import { request } from 'node:https';
 
 const PROPS = 'Directory.Build.props';
 const RE = /(<PackageValidationBaselineVersion[^>]*>)([^<]+)(<\/PackageValidationBaselineVersion>)/;
+const NUGET_INDEX = 'https://api.nuget.org/v3-flatcontainer/queryspec.core/index.json';
 
-let prevTag;
+function fetchPublishedVersions() {
+    return new Promise((resolve, reject) => {
+        const req = request(NUGET_INDEX, (res) => {
+            let body = '';
+            res.on('data', (chunk) => body += chunk);
+            res.on('end', () => {
+                if (res.statusCode !== 200) {
+                    return reject(new Error(`NuGet API returned HTTP ${res.statusCode}`));
+                }
+                try {
+                    resolve(JSON.parse(body).versions);
+                } catch (e) {
+                    reject(e);
+                }
+            });
+        });
+        req.on('error', reject);
+        req.end();
+    });
+}
+
+let tags;
 try {
-    prevTag = execSync('git describe --tags --abbrev=0 --match=v*', { encoding: 'utf8' }).trim();
+    const raw = execSync('git tag -l --sort=-v:refname', { encoding: 'utf8' }).trim();
+    tags = raw ? raw.split('\n').filter(Boolean) : [];
 } catch (err) {
-    console.error('postbump-baseline: no prior v* tag found; skipping baseline update.');
+    console.error('postbump-baseline: git tag failed; skipping baseline update.');
     console.error(err.stderr?.toString() ?? err.message);
     process.exit(0);
 }
 
-if (!/^v\d+\.\d+\.\d+/.test(prevTag)) {
-    console.error(`postbump-baseline: unexpected tag format '${prevTag}'; skipping baseline update.`);
+if (tags.length === 0) {
+    console.log('postbump-baseline: no v* tags found; skipping baseline update.');
     process.exit(0);
 }
-const prevVersion = prevTag.slice(1);
+
+const candidates = tags
+    .filter(t => /^v\d+\.\d+\.\d+/.test(t))
+    .map(t => t.slice(1));
+
+if (candidates.length === 0) {
+    console.log('postbump-baseline: no v* tags with semver format found; skipping baseline update.');
+    process.exit(0);
+}
+
+let published;
+try {
+    published = new Set(await fetchPublishedVersions());
+} catch (err) {
+    console.error(`postbump-baseline: could not query NuGet — ${err.message}`);
+    console.error('Leaving Directory.Build.props unchanged to avoid writing an unverified baseline.');
+    process.exit(0);
+}
+
+const baseline = candidates.find(v => published.has(v));
+
+if (!baseline) {
+    console.log('postbump-baseline: no published version found among candidate tags; leaving Directory.Build.props unchanged.');
+    process.exit(0);
+}
+
+if (candidates[0] !== baseline) {
+    const skipped = candidates.slice(0, candidates.indexOf(baseline));
+    console.log(`postbump-baseline: ${skipped.map(v => 'v' + v).join(', ')} not published on NuGet.org; falling back to v${baseline}`);
+}
 
 const content = readFileSync(PROPS, 'utf8');
 const m = content.match(RE);
@@ -39,11 +97,11 @@ if (!m) {
 }
 
 const oldBaseline = m[2];
-if (oldBaseline === prevVersion) {
-    console.log(`postbump-baseline: baseline already ${prevVersion}, no change.`);
+if (oldBaseline === baseline) {
+    console.log(`postbump-baseline: baseline already ${baseline}, no change.`);
     process.exit(0);
 }
 
-const updated = content.replace(RE, `$1${prevVersion}$3`);
+const updated = content.replace(RE, `$1${baseline}$3`);
 writeFileSync(PROPS, updated);
-console.log(`postbump-baseline: ${PROPS} PackageValidationBaselineVersion ${oldBaseline} -> ${prevVersion}`);
+console.log(`postbump-baseline: ${PROPS} PackageValidationBaselineVersion ${oldBaseline} -> ${baseline}`);

